File: contrib/clojure-package/examples/rnn/src/rnn/test_char_rnn.clj
;;
;; Licensed to the Apache Software Foundation (ASF) under one or more
;; contributor license agreements.  See the NOTICE file distributed with
;; this work for additional information regarding copyright ownership.
;; The ASF licenses this file to You under the Apache License, Version 2.0
;; (the "License"); you may not use this file except in compliance with
;; the License.  You may obtain a copy of the License at
;;
;;    http://www.apache.org/licenses/LICENSE-2.0
;;
;; Unless required by applicable law or agreed to in writing, software
;; distributed under the License is distributed on an "AS IS" BASIS,
;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
;; See the License for the specific language governing permissions and
;; limitations under the License.
;;

(ns rnn.test-char-rnn
  (:require [clojure.string :as string]
            [clojure.java.shell :refer [sh]]
            [rnn.util :as util]
            [rnn.lstm :as lstm]
            [org.apache.clojure-mxnet.context :as context]
            [org.apache.clojure-mxnet.executor :as executor]
            [org.apache.clojure-mxnet.module :as m]
            [org.apache.clojure-mxnet.ndarray :as ndarray]))

(when-not (.exists (clojure.java.io/file "data"))
  (do (println "Retrieving data...") (sh "./get_data.sh")))

(def data-path "data/obama.txt")
(def model-prefix)
(def start-sentence "The joke ")
(def num-hidden 512) ;; hidden unit in LSTM cell
(def num-embed 256) ;; the embedding dim (a char is mapped to 256 dim)
(def num-lstm-layer 3) ;; number of lstm layers

(def vocab (util/build-vocab data-path))

(defn rnn-test [model-prefix epoch-num seq-length random?]
  (let [trained-mod (m/load-checkpoint {:prefix model-prefix :epoch epoch-num})
        trained-arg-params (m/arg-params trained-mod)
        model (lstm/lstm-inference-model {:num-lstm-layer 3
                                          :input-size (inc (count vocab))
                                          :num-label (inc (count vocab))
                                          :num-hidden num-hidden
                                          :num-embed num-embed
                                          :arg-params trained-arg-params})
        input-ndarray (ndarray/zeros [1])
        revert-vocab (util/make-revert-vocab vocab)
        fix-dict (into [""]
                       (mapv #(str (get revert-vocab %))
                             (sort (vals vocab))))
        random-sample random? ;; use this to do random sample or max prob
        ignore-length (count start-sentence)]
    (println "Starter sentence: " start-sentence)
    (println "===")
    (loop [i 0
           new-sentence true
           output start-sentence]
      (if (= seq-length i)
        output
        (do
          (if (<= i (dec ignore-length))
            (util/make-input (get start-sentence i) vocab input-ndarray)
            (util/make-input (last output) vocab input-ndarray))
          (let [prob (ndarray/->vec (lstm/forward model input-ndarray new-sentence))
                next-char (util/make-output prob fix-dict random-sample)]
            (recur (inc i)
                   (if (= "" next-char) true false)
                   (if (< i (dec ignore-length))
                     output
                     (str output next-char)))))))))

(comment

  (rnn-test "data/obama" 75 200 false)
  ;=>"The joke that we can start by the challenges of the American people. The American people have been talking about how to compete with the streets of San Antonio who the courage to come together as one "

  (rnn-test "data/obama" 75 200 true)
  ;=>"The joke before them prepared for five years ago, we only hear a chance to lose our efforts and they made striggling procedural deficit at the city between a politics in the efforts on the Edmund Pett"
)

File: scala-package/examples/src/main/scala/org/apache/mxnetexamples/rnn/TestCharRnn.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.mxnetexamples.rnn

import org.apache.mxnet._
import org.apache.mxnetexamples.InferBase
import org.apache.mxnetexamples.benchmark.CLIParserBase
import org.kohsuke.args4j.{CmdLineParser, Option}
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._

/**
 * Follows the demo, to test the char rnn:
 * https://github.com/dmlc/mxnet/blob/master/example/rnn/char-rnn.ipynb
 */
object TestCharRnn {

  private val logger = LoggerFactory.getLogger(classOf[TrainCharRnn])

  def runInferenceCharRNN(dataPath: String, modelPrefix: String, starterSentence : String): Unit = {
    ResourceScope.using() {
      // The batch size for training
      val batchSize = 32
      // We can support various length input
      // For this problem, we cut each input sentence to length of 129
      // So we only need fix length bucket
      val buckets = List(129)
      // hidden unit in LSTM cell
      val numHidden = 512
      // embedding dimension, which is, map a char to a 256 dim vector
      val numEmbed = 256
      // number of lstm layer
      val numLstmLayer = 3

      // build char vocabluary from input
      val vocab = Utils.buildVocab(dataPath)

      // load from check-point
      val (_, argParams, _) = Model.loadCheckpoint(modelPrefix, 75)

      // build an inference model
      val model = new RnnModel.LSTMInferenceModel(numLstmLayer, vocab.size + 1,
        numHidden = numHidden, numEmbed = numEmbed,
        numLabel = vocab.size + 1, argParams = argParams, dropout = 0.2f)

      // generate a sequence of 1200 chars
      val seqLength = 1200
      val inputNdarray = NDArray.zeros(1)
      val revertVocab = Utils.makeRevertVocab(vocab)

      // Feel free to change the starter sentence
      var output = starterSentence
      val randomSample = true
      var newSentence = true
      val ignoreLength = output.length()

      for (i <- 0 until seqLength) {
        if (i <= ignoreLength - 1) Utils.makeInput(output(i), vocab, inputNdarray)
        else Utils.makeInput(output.takeRight(1)(0), vocab, inputNdarray)
        val prob = model.forward(inputNdarray, newSentence)
        newSentence = false
        val nextChar = Utils.makeOutput(prob, revertVocab, randomSample)
        if (nextChar == "") newSentence = true
        if (i >= ignoreLength) output = output ++ nextChar
      }

      // Let's see what we can learned from char in Obama's speech.
      logger.info(output)
    }
  }

  def main(args: Array[String]): Unit = {
    val stcr = new CLIParser
    val parser: CmdLineParser = new CmdLineParser(stcr)
    try {
      parser.parseArgument(args.toList.asJava)
      assert(stcr.dataPath != null && stcr.modelPrefix != null && stcr.starterSentence != null)
      runInferenceCharRNN(stcr.dataPath, stcr.modelPrefix, stcr.starterSentence)
    } catch {
      case ex: Exception => {
        logger.error(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
      }
    }
  }
}

class CLIParser extends CLIParserBase {
  @Option(name = "--data-path", usage = "the input train data file")
  val dataPath: String = "./data/obama.txt"
  @Option(name = "--model-prefix", usage = "the model prefix")
  val modelPrefix: String = "./model/obama"
  @Option(name = "--starter-sentence", usage = "the starter sentence")
  val starterSentence: String = "The joke"
}

class TestCharRnn(CLIParser: CLIParser) extends InferBase {

  private var vocab : Map[String, Int] = null

  override def loadModel(context: Array[Context], batchInference : Boolean = false): Any = {
    val batchSize = 32
    val buckets = List(129)
    val numHidden = 512
    val numEmbed = 256
    val numLstmLayer = 3
    val (_, argParams, _) = Model.loadCheckpoint(CLIParser.modelPrefix, 75)
    this.vocab = Utils.buildVocab(CLIParser.dataPath)
    var ctx = Context.cpu()
    if (System.getenv().containsKey("SCALA_TEST_ON_GPU") &&
      System.getenv("SCALA_TEST_ON_GPU").toInt == 1) {
      ctx = Context.gpu()
    }
    val model = new RnnModel.LSTMInferenceModel(numLstmLayer, vocab.size + 1,
      numHidden = numHidden, numEmbed = numEmbed,
      numLabel = vocab.size + 1, argParams = argParams, dropout = 0.2f, ctx = ctx)
    model
  }

  override def loadSingleData(): Any = {
    val revertVocab = Utils.makeRevertVocab(vocab)
    revertVocab
  }

  override def runSingleInference(loadedModel: Any, input: Any): Any = {
    val model = loadedModel.asInstanceOf[RnnModel.LSTMInferenceModel]
    val revertVocab = input.asInstanceOf[Map[Int, String]]
    // generate a sequence of 1200 chars
    val seqLength = 1200
    val inputNdarray = NDArray.zeros(1)
    // Feel free to change the starter sentence
    var output = CLIParser.starterSentence
    val randomSample = true
    var newSentence = true
    val ignoreLength = output.length()

    for (i <- 0 until seqLength) {
      if (i <= ignoreLength - 1) Utils.makeInput(output(i), vocab, inputNdarray)
      else Utils.makeInput(output.takeRight(1)(0), vocab, inputNdarray)
      val prob = model.forward(inputNdarray, newSentence)
      newSentence = false
      val nextChar = Utils.makeOutput(prob, revertVocab, randomSample)
      if (nextChar == "") newSentence = true
      if (i >= ignoreLength) output = output ++ nextChar
    }
    output
  }

  override def loadBatchFileList(batchSize: Int): List[Any] = null
  override def loadInputBatch(source: Any): Any = null
  override def runBatchInference(loadedModel: Any, input: Any): Any = null
}
